Sbt plugin section (#237)
higherkindness/compendium: docs/docs/docs/sbt/index.md

---
layout: docs
title: sbt plugin
url: /docs/sbt
---

# What is `sbt-compendium`?

`sbt-compendium` is a plugin that provides a client to communicate with [Compendium](https://github.com/higherkindness/compendium) service. It currently support:

- Retrieval of protocols

To start using `sbt-compendium`, add the following line in `project/plugin.sbt`:

```
addSbtPlugin("io.higherkindness" %% "sbt-compendium" % <version>)

> ```

Reviewer: We can do it in a following PR but `sbt-org-policies` supports replacing versions with the current one on each release. See `[comment]: # (Start Replace)` and `[comment]: # (End Replace)` here https://raw.githubusercontent.com/higherkindness/mu-scala/7f84033fc2de7c6f6a33462470fec70b4ac1cdbf/docs/src/main/docs/reference/modules-artifacts.md

Reviewer 2: I would rather not depend on `sbt-org-policies` so it is easy to migrate out of it. But I think `sbt-mdoc` has a similar feature :thinking: I would look at it in another PR, sounds good?

---
layout: docs
title: sbt plugin
url: /docs/sbt
---

# What is `sbt-compendium`?

`sbt-compendium` is a plugin that provides a client to communicate with [Compendium](https://github.com/higherkindness/compendium) service. It currently support:

> 

Reviewer 2: ```suggestion
```

- Retrieval of protocols

> 

Reviewer 2: ```suggestion
```

---
layout: docs
title: sbt plugin
url: /docs/sbt
---

# What is `sbt-compendium`?

`sbt-compendium` is a plugin that provides you client for
[compendium](https://github.com/higherkindness/compendium). It allows
you to communicate with a compendium service.

# Getting started

Add the following line to your `project/plugin.sbt`:

```mdoc
    addSbtPlugin("io.higherkindness" %% "sbt-compendium" % "0.0.1")

> ```

Reviewer 2: ```suggestion
```
addSbtPlugin("io.higherkindness" %% "sbt-compendium" % <version>)
```
```

The following will also need to be added to your project settings in `build.sbt`:

> ```mdoc

Reviewer 2: ```suggestion
```
```

    .settings(
         compendiumSrcGenProtocolIdentifiers := List(ProtocolAndVersion("supplier",None),ProtocolAndVersion("material",None),ProtocolAndVersion("sale",None)),
         compendiumSrcGenServerHost := "localhost",
         compendiumSrcGenServerPort := 8080,
         sourceGenerators in Compile += Def.task {
           compendiumSrcGenClients.value
         }.taskValue
    )

> ```

Reviewer 2: ```suggestion
```
.settings(
   compendiumSrcGenProtocolIdentifiers := List(ProtocolAndVersion("supplier",None),ProtocolAndVersion("material",None),ProtocolAndVersion("sale",None)),
   compendiumSrcGenServerHost := "localhost",
   compendiumSrcGenServerPort := 8080,
   sourceGenerators in Compile += Def.task {
     compendiumSrcGenClients.value
   }.taskValue
)
```
```

---
layout: docs
title: sbt plugin
url: /docs/sbt
---

# What is `sbt-compendium`?

`sbt-compendium` is a plugin that provides you client for
[compendium](https://github.com/higherkindness/compendium). It allows

> you to communicate with a compendium service.

Reviewer 2: ```suggestion
`sbt-compendium` is a plugin that provides a client to communicate with [Compendium](https://github.com/higherkindness/compendium) service. It currently support:

- Retrieval of protocols
```

## Getting started

> Add the following line to your `project/plugin.sbt`:

Reviewer 2: Merge this in `What is `sbt-compendium` intro

```suggestion
To start using `sbt-compendium`, add the following line in `project/plugin.sbt`:
```

```
addSbtPlugin("io.higherkindness" %% "sbt-compendium" % <version>)
```

The following will also need to be added to your project settings in `build.sbt`:

```
.settings(
  compendiumSrcGenProtocolIdentifiers := List(
    ProtocolAndVersion("supplier",None),
    ProtocolAndVersion("material",None),
    ProtocolAndVersion("sale",None)
  ),
  compendiumSrcGenServerHost := "localhost",
  compendiumSrcGenServerPort := 8080,
  sourceGenerators in Compile += Def.task {
    compendiumSrcGenClients.value
  }.taskValue
)
```

**Note:** These example settings come from

> [compendium-example](https://github.com/higherkindness/compendium-example).

Reviewer 2: Move :arrow_down: to the bottom of the section.

```suggestion
```

The configuration works as follows:

-  `compendiumSrcGenServerHost`: *String*. Url of the compendium server.
   Default value: "localhost"
-  `compendiumSrcGenServerPort`: *Integer*. Port of the compendium
   server. Default value: 47047
-  `compendiumSrcGenFormatSchema`: *IdlName type*. Schema type to
   download. Default value: IdlName.Avro. Currently supported: Avro,
   Proto.
-  `compendiumSrcGenProtocolIdentifiers`: *Seq[ProtocolAndVersion]*.
   Protocol identifiers to be retrieved from the compendium server.
   `ProtocolAndVersion` provides two values: `name` (mandatory) that
   corresponds with the identifier used to store the protocol and

>    `version` (optional). Default: Nil

Reviewer 2: ```suggestion
## Available settings

| Setting | Type | Description | Default value |
|---|---|---|---|
| `compendiumSrcGenServerHost` | _String_ | Compendium server host | `localhost` |
| `compendiumSrcGenServerPort` | _Integer_ | Compendium server port | `47047` |
| `compendiumSrcGenFormatSchema` | _IdlName_ | Schema type to download | `IdlName.Avro` |
| `compendiumSrcGenProtocolIdentifiers` | _Seq[ProtocolAndVersion]_ | Protocol identifiers to retrieve from compendium. `ProtocolAndVersion` provides two values: `name` (mandatory) that corresponds with the identifier used to store the protocol and `version` (optional) | `Nil` | 
```

## How to use it

Once you have set up the configuration, `sbt-compendium` will, during compilation,
get the protocols from the compendium service, validate them and create the
proper Scala classes in `target/scala-2.12/src_managed`.

### What if I want to save a class on compendium?

`sbt-compendium` is designed to retrieve protocols but not to save them If
you want to save protocols you'll need to make an http call to

```
[compendium host and port]/v0/protocol/[identifier]?idlName=[format]
```

with the body

```
{
  "raw" : [protocol string formatted]
}

> ```

Reviewer 2: IMHO, I do not think this should be explain in `sbt-compendium` section, since right now it is out of the scope of this plugin for now.

```suggestion
```

Instead, I would place the `##Example` section here :wink: (please, you need to add the following in yourself since I could not use nested markdown code sections in `suggestions`. 

---

## Example

And example of `sbt-compendium` usage is available at [compendium-example](https://github.com/higherkindness/compendium-example). It looks like:

```
.settings(
  compendiumSrcGenProtocolIdentifiers := List(
    ProtocolAndVersion("supplier",None),
    ProtocolAndVersion("material",None),
    ProtocolAndVersion("sale",None)
  ),
  compendiumSrcGenServerHost := "localhost",
  compendiumSrcGenServerPort := 8080,
  sourceGenerators in Compile += Def.task {
    compendiumSrcGenClients.value
  }.taskValue
)
```
